### PR TITLE
Use autoencoder for inputs for hierarchical xc

### PR DESCRIPTION
**Pull request by sipian**
Ref #5

### CODE REVIEW

File: train_hierarchialXC.py [hunk 1]
```diff
@@ -23,8 +28,14 @@
 # training configuration arguments
 parser.add_argument('--seed', type=int, default=None,
                     help='Manually set the seed for the experiments for reproducibility')
-parser.add_argument('--plot', action='store_true',
-                    help='Option to plot the loss variation over iterations')
+parser.add_argument('--interval', type=int, default=-1,
+                    help='Interval between two status updates on training')
+parser.add_argument('--input_ae_dim', type=int, default=500,
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Set the default to -1. If -1, then no dimensionality reduction is done on the inputs.

File: train_hierarchialXC.py [hunk 2]
```diff
@@ -23,8 +28,14 @@
...
+                    help='Dimensionality of encoding in the input encoder')
+parser.add_argument('--input_ae_factor', type=float, default=0.25,
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Irrelevant.

Comment (vishwakftw) [reply train_hierarchialXC.py]: Let me elaborate, I was under the impression that we'll set this parameter permanently.

File: train_hierarchialXC.py [hunk 3]
```diff
@@ -42,15 +53,47 @@
...
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Train your model ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-my_hierarchical_XC.train(loader, multiclass.OneVsRestClassifier, estimator=svm.SVC())
-predictions = my_hierarchical_XC.predict(loader.get_data()[0])
+all_iters = 0
+start_time = time.time()
+cur_no = 0
+ae = GenericAutoencoder(loader.get_features().shape[1], args.input_ae_factor, args.input_ae_dim)
+
+for x, y in iter(data_loader):
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: make this `(x, _)` because we don't care about `y`.

File: train_hierarchialXC.py [hunk 4]
```diff
@@ -42,15 +53,47 @@
...
+
+    y = np.squeeze(y.numpy(), 1)
+    y = csr_matrix(y)
+
+    all_iters += 1
+    if all_iters % args.interval == 0:
+        print("Iter #{} :: {} / {} - INP_REC_LOSS : {}"
+              .format(all_iters, cur_no, len_loader, round(loss_inp_rec.item(), 5)))
+
+    my_hierarchical_XC.train(inp_ae_fp, y, multiclass.OneVsRestClassifier,
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: No, got the procedure wrong. We do this training, and then train the whole bunch and the encodings.

Comment (GoodDeeds) [inline train_hierarchialXC.py]: I think entire dataset needs to be passed to `train`, else `len(loader)` number of trees will be created.

Comment (vishwakftw) [reply train_hierarchialXC.py]: Or at least that was what I had in mind.

File: train_hierarchialXC.py [hunk 5]
```diff
@@ -42,15 +53,47 @@
...
+cur_no = 0
+ae = GenericAutoencoder(loader.get_features().shape[1], args.input_ae_factor, args.input_ae_dim)
+
+for x, y in iter(data_loader):
+
+    cur_no += x.size(0)
+
+    inp_ae_fp = ae.forward(x.float())
+    loss_inp_rec = F.mse_loss(inp_ae_fp, x.float())
+    inp_ae_fp = np.squeeze(inp_ae_fp.detach().numpy(), 1)
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: not required.

File: train_hierarchialXC.py [hunk 6]
```diff
@@ -42,15 +53,47 @@
...
+    inp_ae_fp = csr_matrix(inp_ae_fp)
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: not required.

File: train_hierarchialXC.py [hunk 7]
```diff
@@ -42,15 +53,47 @@
...
+
+    y = np.squeeze(y.numpy(), 1)
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: not required.

File: train_hierarchialXC.py [hunk 8]
```diff
@@ -42,15 +53,47 @@
...
+    y = csr_matrix(y)
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: not required.

File: train_hierarchialXC.py [hunk 9]
```diff
@@ -42,15 +53,47 @@
...
+
+    my_hierarchical_XC.train(inp_ae_fp, y, multiclass.OneVsRestClassifier,
+                             estimator=svm.SVC(), n_jobs=-1)
+
+print("--- Completed Training in %.5f seconds ---" % (time.time() - start_time))
 
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Calculate the metrics ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+start_time = time.time()
+predictions = my_hierarchical_XC.predict(loader.get_data()[0])
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: `my_hierarchicalXC.predict()` goes here.

**Comment by vishwakftw** [comment]
@sipian status?

File: train_hierarchialXC.py (hunk 1 above)
Comment (sipian) [reply train_hierarchialXC.py]: Setting this to -1, is leading OOM killed errors.

File: train_hierarchialXC.py [hunk 10]
```diff
@@ -62,36 +62,40 @@
...
+my_hierarchical_XC.train(x, loader.get_classes(), multiclass.OneVsRestClassifier,
+                         estimator=svm.SVC(), n_jobs=-1)
 
 print("--- Completed Training in %.5f seconds ---" % (time.time() - start_time))
 
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Calculate the metrics ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 start_time = time.time()
-predictions = my_hierarchical_XC.predict(loader.get_data()[0])
+predictions = my_hierarchical_XC.predict(loader.get_features())
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: No, you are wrong here. We have to pass the autoencoded inputs here too.

File: train_hierarchialXC.py (hunk 1 above)
Comment (vishwakftw) [reply train_hierarchialXC.py]: No, I am saying is let the auto-encoding be done based on requirements. If auto-encoding is required, we give a positive val, else we leave it.

Comment (vishwakftw) [reply train_hierarchialXC.py]: Mention the default value in the description.

File: train_hierarchialXC.py [hunk 11]
```diff
@@ -62,36 +62,40 @@
...
+        all_iters += 1
+        if all_iters % args.interval == 0:
+            print("{} / {} :: {} / {} - INP_REC_LOSS : {}\t"
+                  .format(epoch, args.epochs, cur_no, len_loader, round(loss_inp_rec.item(), 5)))
+        INP_REC_LOSS.append(loss_inp_rec.item())
 
-    y = np.squeeze(y.numpy(), 1)
-    y = csr_matrix(y)
+x = loader.get_features().toarray()
+x = ae.forward(torch.from_numpy(x).float())
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Do this in batches, and stack the data. Something like this:
```
def generate_encodings(data_loader, ae):
    vals = []
    for x, _ in iter(data_loader):
        vals.append(ae(x))
    vals = torch.cat(vals, dim=0)
    return vals
```

Use this function to generate encodings for train and test.

File: train_hierarchialXC.py [hunk 12]
```diff
@@ -61,41 +61,42 @@
...
+            all_iters += 1
+            if all_iters % args.interval == 0:
+                print("{} / {} :: {} / {} - INP_REC_LOSS : {}\t"
+                      .format(epoch, args.epochs, cur_no, len_loader, round(loss_inp_rec.item(), 5)))
+            INP_REC_LOSS.append(loss_inp_rec.item())
 
-my_hierarchical_XC.train(x, loader.get_classes(), multiclass.OneVsRestClassifier,
-                         estimator=svm.SVC(), n_jobs=-1)
+    input = input.toarray()
+    input = ae.forward(torch.from_numpy(input).float())
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Refer to my old comment please. For large datasets, a full forward is not possible.

And also remove the `.float()` calls. This has been fixed in master.

File: train_hierarchialXC.py [hunk 13]
```diff
@@ -15,6 +15,18 @@
 import numpy as np
 import torch.nn.functional as F
 
+def generate_encodings(data_loader, ae):
+    vals = []
+
+    for x, _ in iter(data_loader):
+        vals.append(ae(x.float()))
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Is the float issue still there?

File: train_hierarchialXC.py [hunk 14]
```diff
@@ -15,6 +15,18 @@
...
+    vals = torch.cat(vals, dim=0)
+
+    vals = vals.detach().numpy()
+    vals = np.squeeze(vals, 1)
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Why squeeze?

File: train_hierarchialXC.py (hunk 13 above)
Comment (sipian) [reply train_hierarchialXC.py]: I am fixing this

File: train_hierarchialXC.py (hunk 14 above)
Comment (sipian) [reply train_hierarchialXC.py]: dimensions of vals is `[no_of_datapoints,1,1836]`

Comment (vishwakftw) [reply train_hierarchialXC.py]: That’s funny. I thought I fixed this.

If so, let it be like this.

Comment (vishwakftw) [reply train_hierarchialXC.py]: Or rather squeeze the inputs to be `batch size x dims`. Do this, this is better.

**Comment by vishwakftw** [comment]
Flake failing.

File: train_hierarchialXC.py (hunk 14 above)
Comment (vishwakftw) [reply train_hierarchialXC.py]: IIRC, this issue will not happen if you use the data loader. I remember fixing the logic in the `__getitem__` method. So the squeeze should not be necessary. Consult the code in train_neuralXC for this

File: train_hierarchialXC.py [hunk 15]
```diff
@@ -58,6 +69,7 @@ def generate_encodings(data_loader, ae):
 # parse the arguments
 args = parser.parse_args()
 
+print(" a ad ", args.save_model)
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Heh

Comment (sipian) [reply train_hierarchialXC.py]: oops. can i force push :) ?

Comment (vishwakftw) [reply train_hierarchialXC.py]: Push changes with the removal of the squeeze method

File: train_hierarchialXC.py (hunk 14 above)
Comment (sipian) [reply train_hierarchialXC.py]: Since this is for vals, I didn't quite get how do you plan to use data_loader for this

Comment (vishwakftw) [reply train_hierarchialXC.py]: `vals` is the concatenation of forward passes over the inputs to the network, which are items in the data loader.

File: train_hierarchialXC.py [hunk 16]
```diff
@@ -29,7 +29,7 @@ def generate_encodings(data_loader, ae):
     """
     vals = []
     for x, _ in iter(data_loader):
-        x =  x.view(-1, x.size()[2])
+        x = x.view(-1, x.size()[2])
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Why so complicated?

`x = x.squeeze(1)` should have worked just fine.

File: train_hierarchialXC.py [hunk 17]
```diff
@@ -42,15 +77,50 @@
...
+    INP_REC_LOSS = []
+
+    for epoch in range(args.epochs):
+        cur_no = 0
+        for x, _ in iter(data_loader):
+            cur_no += x.size(0)
+
+            inp_ae_fp = ae.forward(x)
+            loss_inp_rec = F.mse_loss(inp_ae_fp, x)
+
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Dude, you forgot `loss_inp_rec.backward()`.

File: train_hierarchialXC.py [hunk 18]
```diff
@@ -42,15 +77,50 @@
...
+    cur_no = 0
+    ae = GenericAutoencoder(input.shape[1], 0.25, args.input_ae_dim)
+
+    INP_REC_LOSS = []
+
+    for epoch in range(args.epochs):
+        cur_no = 0
+        for x, _ in iter(data_loader):
+            cur_no += x.size(0)
+
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: You forgot `optimizer.zero_grad()`.

File: train_hierarchialXC.py [hunk 19]
```diff
@@ -23,8 +51,15 @@
...
-                    help='Option to plot the loss variation over iterations')
+parser.add_argument('--interval', type=int, default=-1,
+                    help='Interval between two status updates on training')
+parser.add_argument('--input_ae_dim', type=int, default=-1,
+                    help="""Output dimensions of the input encoder.
+                    By default, this is -1, meaning no encoding is done""")
+parser.add_argument('--epochs', type=int, default=10,
+                    help='Number of epochs to train the autoencoder for')
+parser.add_argument('--batch_size', type=int, default=64,
+                    help='Batch size for training')
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Optimizer options?

File: train_hierarchialXC.py [hunk 20]
```diff
@@ -42,15 +77,50 @@
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Dataloader initialization ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Optimizer?

**Comment by vishwakftw** [comment]
This has to be squashed and merged :wink:

File: train_hierarchialXC.py (hunk 20 above)
Comment (sipian) [reply train_hierarchialXC.py]: Should I create a complete `my_neural_XC` for the parameters as in [this](https://github.com/vishwakftw/extreme-classification/blob/master/train_neuralxc.py#L116)?

Comment (sipian) [reply train_hierarchialXC.py]: apologizes for all the childish questions. :)

Comment (vishwakftw) [reply train_hierarchialXC.py]: No no. This would how the optimizer would look like:
```
optimizer = getattr(torch.optim, opt_options['name'])(ae.parameters(),
**opt_options['args'])
```

File: train_hierarchialXC.py [hunk 21]
```diff
@@ -21,10 +50,23 @@
...
+                    help='Number of epochs to train the autoencoder for')
+parser.add_argument('--interval', type=int, default=-1,
+                    help='Interval between two status updates on training')
+parser.add_argument('--input_ae_dim', type=int, default=-1,
+                    help="""Output dimensions of the input encoder.
+                    By default, this is -1, meaning no encoding is done""")
+
+# optimizer arguments
+parser.add_argument('--optimizer_cfg', type=str, required=True,
+                    help='Optimizer configuration in YAML format for NeuralXC model')
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: s/NeuralXC model/Autoencoder

File: train_hierarchialXC.py [hunk 22]
```diff
@@ -1,14 +1,43 @@
 from datetime import datetime
 from argparse import ArgumentParser
+from scipy.sparse import csr_matrix
 from sklearn import multiclass, svm
 from sklearn.externals import joblib
 
 from extreme_classification.loaders import LibSVMLoader
 from extreme_classification.hierarchicalXC import HierarchicalXC
 from extreme_classification.metrics import precision_at_k, ndcg_score_at_k
+from extreme_classification.autoencoders import GenericAutoencoder
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: move up.

File: train_hierarchialXC.py [hunk 23]
```diff
@@ -1,14 +1,43 @@
...
+        Reconstructed input as a sparse matrix.
+    """
+    vals = []
+    for x, _ in iter(data_loader):
+        x = x.to(device=cur_device, dtype=torch.float)
+        x = x.squeeze(1)
+        vals.append(ae(x))
+
+    vals = torch.cat(vals, dim=0)
+    vals = vals.detach().numpy()
```
Comment (vishwakftw) [inline train_hierarchialXC.py]: Copy to CPU.

Comment (vishwakftw) [reply train_hierarchialXC.py]: vals = vals.detach().cpu().numpy()
